mender-update/standalone/standalone.cpp

```diff
@@ -0,0 +1,516 @@
+		return {Result::FailedNothingDone, payload.error()};
+	}
+
+	cout << "Installing artifact..." << endl;
+
+	auto err = update_module.Download(payload.value());
+	if (err != error::NoError) {
+		err = err.FollowedBy(update_module.Cleanup());
+		err = err.FollowedBy(RemoveStandaloneData(main_context.GetMenderStoreDB()));
+		return {Result::FailedNothingDone, err};
+	}
+
+	err = update_module.ArtifactInstall();
+	if (err != error::NoError) {
+		log::Error("Installation failed: " + err.String());
+		return InstallationFailureHandler(main_context, data, update_module);
+	}
+
+	auto reboot = update_module.NeedsReboot();
+	if (!reboot) {
+		log::Error("Could not query for reboot: " + reboot.error().String());
+		return InstallationFailureHandler(main_context, data, update_module);
+	}
+
+	auto rollback_support = update_module.SupportsRollback();
+	if (!rollback_support) {
+		log::Error("Could not query for rollback support: " + rollback_support.error().String());
+		return InstallationFailureHandler(main_context, data, update_module);
+	}
+
+	if (rollback_support.value()) {
+		if (reboot.value() != update_module::RebootAction::No) {
+			return {Result::InstalledRebootRequired, error::NoError};
+		} else {
+			return {Result::Installed, error::NoError};
+		}
+	}
+
+	cout << "Update Module doesn't support rollback. Committing immediately." << endl;
+
+	auto result = DoCommit(main_context, data, update_module);
+	if (result.result == Result::Committed) {
+		if (reboot.value() != update_module::RebootAction::No) {
+			result.result = Result::InstalledAndCommittedRebootRequired;
+		} else {
+			result.result = Result::InstalledAndCommitted;
+		}
+	}
+	return result;
+}
+
+ResultAndError DoCommit(
+	context::MenderContext &main_context,
+	StandaloneData &data,
+	update_module::UpdateModule &update_module) {
+	auto err = update_module.ArtifactCommit();
+	if (err != error::NoError) {
+		log::Error("Commit failed: " + err.String());
+		return InstallationFailureHandler(main_context, data, update_module);
+	}
+
+	auto result = Result::Committed;
+	error::Error return_err;
+
+	err = update_module.Cleanup();
+	if (err != error::NoError) {
+		result = Result::InstalledButFailedInPostCommit;
+		return_err = return_err.FollowedBy(err);
+	}
+
+	err = main_context.CommitArtifactData(
+		data.artifact_name,
+		data.artifact_group,
+		data.artifact_provides,
+		data.artifact_clears_provides,
+		[](database::Transaction &txn) {
+			return txn.Remove(context::MenderContext::standalone_state_key);
+		});
+	if (err != error::NoError) {
+		result = Result::InstalledButFailedInPostCommit;
+		return_err = return_err.FollowedBy(err);
+	}
+
+	return {result, return_err};
+}
+
+ResultAndError DoRollback(
+	context::MenderContext &main_context,
+	StandaloneData &data,
+	update_module::UpdateModule &update_module) {
+	auto exp_rollback_support = update_module.SupportsRollback();
+	if (!exp_rollback_support) {
+		return {Result::NoRollback, exp_rollback_support.error()};
+	} else {
```

Comment:
Done. I kept one unnecessary else for the `RolledBack`/`NoRollback` pair. Since neither is a "preferred" path, it makes sense to keep them at the same level.